cardano-crypto-tests/test/Main.hs

```diff
@@ -9,21 +9,29 @@ import qualified Test.Crypto.Vector.Secp256k1DSIGN
 import Test.Tasty (TestTree, adjustOption, testGroup, defaultMain)
 import Test.Tasty.QuickCheck (QuickCheckTests (QuickCheckTests))
 import Cardano.Crypto.Libsodium (sodiumInit)
+import Test.Crypto.Util (Lock, mkLock)
 
 main :: IO ()
 main = do
-    sodiumInit
-    defaultMain tests
+  sodiumInit
 
-tests :: TestTree
-tests =
+  -- This lock is used to prevent tests that use mlocking from running
+  -- concurrently. Concurrent execution of these tests can cause the testsuite
+  -- to exhaust mlock quota; but each individual test on its own should be
+  -- fine.
+  mlockLock <- mkLock
+
+  defaultMain (tests mlockLock)
+
+tests :: Lock -> TestTree
+tests mlockLock =
   -- The default QuickCheck test count is 100. This is too few to catch
   -- anything, so we set a minimum of 1000.
   adjustOption (\(QuickCheckTests i) -> QuickCheckTests $ max i 1000) .
     testGroup "cardano-crypto-class" $
-      [ Test.Crypto.DSIGN.tests
-      , Test.Crypto.Hash.tests
-      , Test.Crypto.KES.tests
+      [ Test.Crypto.DSIGN.tests mlockLock
+      , Test.Crypto.Hash.tests mlockLock
+      , Test.Crypto.KES.tests mlockLock
```

Comment:
Since we can't run those tests concurrently I believe it should be enough to make those tests run sequentially by changing the setting, instead of using a lock: 
```suggestion
      [ localOption (NumThreads 1) $ testGroup $ "Sequentially run tests" [
          Test.Crypto.DSIGN.tests
        , Test.Crypto.Hash.tests
        , Test.Crypto.KES.tests
        ]
```

Reply:
Technically, it's just the section between acquiring and releasing mlocked memory that cannot run concurrently; the rest of the tests, i.e., generating and selecting inputs (which is done in unprotected memory due to `Arbitrary` being a pure interface), and turning results into human-readable output, can be allowed to run concurrently just fine.

Additionally, there are some tests that can run concurrently just fine, because they don't touch mlocked memory at all. E.g. the "pure" flavors of `Hash` and `DSIGN`), as well as tests that allocate mlocked memory only for part of their lifetime (e.g. tests for KES verification, which allocate some mlocked memory to make a sign key, then they will derive a verification key and sign something, *discard the sign key*, and then do the actual testing, like verifying the signature. Only the part until "discard the sign key" needs to be protected, the rest can, and should, run concurrently just fine.

Forcing these subtrees to run on a single thread will of course work, but I'd expect the test suite to take even longer to run.

Reply:
We don't have to force all subtrees to run sequentially. This was just a depiction on how it can be done, this local option can be supplied only at places were sequential running of tests is needed. So make a few groups of  tests, which use the mlocked memory, into a tree and mark them sequential. There is still need to for passing around a lock manually.

Reply:
As discussed on Slack: `localOptions (NumThreads 1)` is a no-op in Tasty, so this won't work, and the explicit lock is the only way to force some, but not all, tests to run sequentially, so we should leave it as-is.